### PR TITLE
Libnpg include dirs fixed

### PR DESCRIPTION
**Pull request by czarny247**
## Checklist

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] I used the same coding style as the surrounding code 
- [x] I have tested the proposed changes
- [x] I extended the documentation, if necessary
- [x] The proposed change builds also on our build bots (check after submission)

### CODE REVIEW

**Comment by czarny247** [comment]
@lynxlynxlynx Could you please kill this action https://github.com/gemrb/gemrb/actions/runs/8053767452 ?
It failed to finish somehow.

**Comment by lynxlynxlynx** [comment]
Thanks, this was annoying. Didn't realize we hardcode the version.

File: .github/workflows/builder.yml [hunk 1]
```diff
@@ -20,6 +20,9 @@ jobs:
       run: |
         brew install sdl2 openal-soft sdl2_mixer libpng libvorbis sccache googletest
         brew reinstall libpng
+        brew --cellar libpng
+        brew list --versions libpng
+        brew list --versions libpng | egrep -o '[0-9]+\.[0-9]+\.[0-9]+'
```
Comment (lynxlynxlynx) [inline .github/workflows/builder.yml:25]: These don't hurt, but now the pointless reinstall can be removed.

File: cmake/FindDependencies.cmake [hunk 2]
```diff
@@ -42,8 +42,10 @@ ENDIF()
 
 IF(USE_PNG)
 	IF(APPLE AND DEFINED ENV{GITHUB_SHA})
-		SET(PNG_INCLUDE_DIRS "/usr/local/Cellar/libpng/1.6.40/include")
-		SET(PNG_LIBRARIES "/usr/local/Cellar/libpng/1.6.40/lib/libpng16.16.dylib")
+		#brew --cellar libpng -> /usr/local/Cellar/libpng
+		#brew list --versions libpng -> libpng 1.6.42
+		SET(PNG_INCLUDE_DIRS "/usr/local/Cellar/libpng/1.6.42/include")
+		SET(PNG_LIBRARIES "/usr/local/Cellar/libpng/1.6.42/lib/libpng16.16.dylib")
```
Comment (lynxlynxlynx) [inline cmake/FindDependencies.cmake:48]: This way we're just kicking the ball down the line. Ideally we could just use globbing, but I'm sceptical that would work.

find_path can be used to locate the dylib, since that name doesn't change:
https://cmake.org/cmake/help/latest/command/find_path.html
(I suggest using NO_DEFAULT_PATH for greater robustness)

... and then you can append "../../include" and bam, we also have the value for the include dir.

File: .github/workflows/builder.yml (hunk 1 above)
Comment (czarny247) [reply .github/workflows/builder.yml]: Removed

File: cmake/DeduceLibpngPath.cmake [hunk 3]
```diff
@@ -0,0 +1,17 @@
+FUNCTION(DEDUCE_LIBPNG_PATH)
```
Comment (lynxlynxlynx) [inline cmake/DeduceLibpngPath.cmake:1]: this is platform-specfic, so should be in platforms/... instead.

File: cmake/DeduceLibpngPath.cmake [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+	execute_process(
+		COMMAND brew list --versions libpng
+		OUTPUT_VARIABLE LIBPNG_WITH_VERSION
+		OUTPUT_STRIP_TRAILING_WHITESPACE)
+
+	set(regex_pattern "([0-9]+\\.[0-9]+\\.[0-9]+)")
+	string(REGEX MATCH "${regex_pattern}" LIBPNG_VERSION "${LIBPNG_WITH_VERSION}")
```
Comment (lynxlynxlynx) [inline cmake/DeduceLibpngPath.cmake:8]: what happens if brew returns more than one version?

File: cmake/FindDependencies.cmake [hunk 5]
```diff
@@ -42,8 +42,15 @@ ENDIF()
 
 IF(USE_PNG)
 	IF(APPLE AND DEFINED ENV{GITHUB_SHA})
-		SET(PNG_INCLUDE_DIRS "/usr/local/Cellar/libpng/1.6.40/include")
-		SET(PNG_LIBRARIES "/usr/local/Cellar/libpng/1.6.40/lib/libpng16.16.dylib")
+		INCLUDE(DeduceLibpngPath)
+		DEDUCE_LIBPNG_PATH()
+
+		if("${LIBPNG_PATH}" STREQUAL "")
```
Comment (lynxlynxlynx) [inline cmake/FindDependencies.cmake:48]: I'm pretty sure cmake treats empty strings as falsy, so this can be simplified.

File: cmake/FindDependencies.cmake [hunk 6]
```diff
@@ -42,8 +42,15 @@ ENDIF()
...
+			message(ERROR "Cannot deduce LIBPNG_PATH!")
```
Comment (lynxlynxlynx) [inline cmake/FindDependencies.cmake:49]: If this happens we shouldn't execute the following lines.

File: cmake/DeduceLibpngPath.cmake (hunk 4 above)
Comment (lynxlynxlynx) [reply cmake/DeduceLibpngPath.cmake:8]: `brew --prefix libpng` probably prints the path directly.

**Comment by czarny247** [comment]
> Thanks, this was annoying. Didn't realize we hardcode the version.

Me too, I started with workaround until realized it :D

File: cmake/FindDependencies.cmake (hunk 2 above)
Comment (czarny247) [reply cmake/FindDependencies.cmake]: I failed making it with `find_path`. It is not so obvious to me and hard to test without macOS locally.

File: cmake/DeduceLibpngPath.cmake (hunk 4 above)
Comment (lynxlynxlynx) [reply cmake/DeduceLibpngPath.cmake]: It's a symlink to the actual location, so it should resolve just fine.

File: CMakeLists.txt [hunk 7]
```diff
@@ -1,12 +1,23 @@
 CMAKE_MINIMUM_REQUIRED(VERSION 3.11)
 
+SET(CMAKE_PROJECT_INCLUDE_BEFORE
+	"${CMAKE_SOURCE_DIR}/cmake/IncludeBeforeProject.cmake"
+)
+
+PROJECT(gemrb)
+
```
Comment (lynxlynxlynx) [inline CMakeLists.txt:8]: why did you move this around?

File: CMakeLists.txt [hunk 8]
```diff
@@ -1,12 +1,23 @@
...
 SET(CMAKE_CXX_STANDARD 14)
 
 LIST(APPEND CMAKE_MODULE_PATH
-	"${CMAKE_SOURCE_DIR}/cmake"
+	"${CMAKE_CURRENT_SOURCE_DIR}/cmake"
 	"${CMAKE_CURRENT_SOURCE_DIR}/cmake/modules/"
 )
 
+IF(CMAKE_SYSTEM_NAME STREQUAL "Darwin")
```
Comment (lynxlynxlynx) [inline CMakeLists.txt:16]: it would be cleaner to test APPLE here, since that's what we use everywhere else.

File: CMakeLists.txt (hunk 7 above)
Comment (czarny247) [reply CMakeLists.txt:8]: 1. Note mentioned here: https://cmake.org/cmake/help/latest/command/project.html#usage
2. `CMAKE_SYSTEM_NAME` is empty before `project` call so i couldn't set this properly:

```
IF(CMAKE_SYSTEM_NAME STREQUAL "Darwin")
	LIST(APPEND CMAKE_MODULE_PATH
		"${CMAKE_CURRENT_SOURCE_DIR}/platforms/apple/cmake")
ENDIF()
```

File: cmake/Helpers.cmake [hunk 9]
```diff
@@ -515,9 +515,9 @@ FUNCTION(CONFIGURE_APPLE_SPECIFICS)
 	FIND_LIBRARY(COCOA_LIBRARY_PATH Cocoa)
 	FIND_LIBRARY(COREFOUNDATION_LIBRARY CoreFoundation)
 	# for objective-c++
-	STRING(APPEND CMAKE_CXX_FLAGS " -DTARGET_OS_MAC")
+	SET(CMAKE_CXX_FLAGS ${CMAKE_CXX_FLAGS} " -DTARGET_OS_MAC")
 	# for pure objective-c
-	STRING(APPEND CMAKE_C_FLAGS " -x objective-c -DTARGET_OS_MAC")
+	SET(CMAKE_C_FLAGS ${CMAKE_C_FLAGS} " -x objective-c -DTARGET_OS_MAC")
```
Comment (lynxlynxlynx) [inline cmake/Helpers.cmake:520]: How are these two things different?
Anyway, since I expect this to fail, the cmake way to add definitions is https://cmake.org/cmake/help/latest/command/add_compile_definitions.html#command:add_compile_definitions

File: CMakeLists.txt (hunk 7 above)
Comment (lynxlynxlynx) [reply CMakeLists.txt:8]: Your own CACHE_BUILD_TYPE says it should happen before the project call, so it needs to be moved if so, now that it's ran later.

I see now that CMAKE_PROJECT_INCLUDE_BEFORE requires 3.15, so please update the minimum version as well.

**Comment by lynxlynxlynx** [comment]
I see you're struggling with the CFLAGS changes. Considering a bunch of the changes in this PR were removed via simplification, I suggest you check if the previous simple string appending still works instead.

**Comment by czarny247** [comment]
> I see you're struggling with the CFLAGS changes. Considering a bunch of the changes in this PR were removed via simplification, I suggest you check if the previous simple string appending still works instead.

I made a search trough the github and realized how it should be done. It works now ;)

**Comment by lynxlynxlynx** [comment]
Hmm. But does the semicolon actually work? I know it compiles, but doesn't the compiler then get a mangled flag?

Anyway, there might be a better way still — perhaps it works without those two lines, since we do reference the files for compilation: https://cmake.org/cmake/help/v3.16/release/3.16.html#languages

**Comment by czarny247** [comment]
> Hmm. But does the semicolon actually work? I know it compiles, but doesn't the compiler then get a mangled flag?
> 
> Anyway, there might be a better way still — perhaps it works without those two lines, since we do reference the files for compilation: https://cmake.org/cmake/help/v3.16/release/3.16.html#languages

I simplified it, take a look at the main CMakeLists.txt

**Comment by lynxlynxlynx** [comment]
Yeah that's much nicer, thanks. Just the minimum version bump is missing.

**Comment by czarny247** [comment]
> Yeah that's much nicer, thanks. Just the minimum version bump is missing.

Done ;)

**Comment by lynxlynxlynx** [comment]
argh, it's 3.16 — remember the obj-c release notes.

**Comment by lynxlynxlynx** [comment]
Thanks!
